src/dodal/devices/areadetector/adaravis.py

```diff
@@ -1,99 +1,186 @@
-            self.cam.trigger_mode: "Off",
-            **self.stage_sigs,  # type: ignore
-        }
-
-        # Settings to apply when priming plugins during pre-stage
-        self._priming_settings = {
-            self.hdf.enable: 1,
-            self.hdf.nd_array_port: self.cam.port_name.get(),
-            self.cam.array_callbacks: 1,
-            self.cam.image_mode: "Single",
-            self.cam.trigger_mode: "Off",
-            # Take the quickest possible frame
-            self.cam.acquire_time: 6.3e-05,
-            self.cam.acquire_period: 0.003,
-        }
-
-        # Signals that control driver and hdf writer should be put_complete to
-        # avoid race conditions during priming
-        for signal in set(self.stage_sigs.keys()).union(
-            set(self._priming_settings.keys())
-        ):
-            if isinstance(signal, EpicsSignal):
-                signal.put_complete = True
-        self.cam.acquire.put_complete = True
-
-    def stage(self, *args, **kwargs):
-        # We have to manually set the acquire period bcause the EPICS driver
-        # doesn't do it for us. If acquire time is a staged signal, we use the
-        # stage value to calculate the acquire period, otherwise we perform
-        # a caget and use the current acquire time.
-        if self.cam.acquire_time in self.stage_sigs:
-            acquire_time = self.stage_sigs[self.cam.acquire_time]
+import asyncio
+import csv
+from enum import Enum
+from pathlib import Path
+from typing import Literal, Optional, get_args
+
+from bluesky.protocols import HasHints, Hints
+from ophyd_async.core import (
+    DEFAULT_TIMEOUT,
+    AsyncStatus,
+    DetectorControl,
+    DetectorTrigger,
+    DirectoryProvider,
+    StandardDetector,
+    set_and_wait_for_value,
+)
+from ophyd_async.epics.areadetector.drivers import ADBase, ADBaseShapeProvider
+from ophyd_async.epics.areadetector.utils import (
+    ImageMode,
+    ad_r,
+    ad_rw,
+    stop_busy_record,
+)
+from ophyd_async.epics.areadetector.writers import HDFWriter, NDFileHDF, NDPluginStats
+
+
+class ADAravisTriggerMode(str, Enum):
+    """GigEVision GenICAM standard: on=externally triggered"""
+
+    on = "On"
+    off = "Off"
+
+
+class ADAravisTriggerSource(str, Enum):
+    # While not all enum elements may be physically supported by the hardware,
+    # DB record templates suggest they are valid options for underlying record
+    # cite: https://github.com/areaDetector/ADGenICam/tree/master/GenICamApp/Db
+    # (e.g. Mako G-125B TriggerSource has Line1-4)
+    freerun = "Freerun"
+    fixed_rate = "FixedRate"
+    software = "Software"
+    action_0 = "Action0"
+    action_1 = "Action1"
+    # Externally triggered on GPIO N
+    line_1 = "Line1"
+    line_2 = "Line2"
+    line_3 = "Line3"
+    line_4 = "Line4"
+
+
+_supported_external_triggers = (
+    DetectorTrigger.constant_gate,
+    DetectorTrigger.edge_trigger,
+)
+_csv_path = Path(__file__).parent / "adaravis_deadtime.csv"
```

Comment:
I was specifically trying to avoid that (mostly when thinking about every single supported camera, now that I've cut down to just the ones in use at Diamond I'm not as opposed). It's only ~30 entries, but there's got to be a good way of handling these configuration/lookup tables/magic numbers (delete as cynicism demands) generally...

Reply:
I would say in general there are three options for configuration/lookup tables/magic numbers:
1. Package with dodal, as you've done here
2. Package with IOC and make available via a PV
3. Put in some other database that can be queried at runtime

My personal favourite is 2. because it's the most standard thing you can do. Possibly worth asking @coretl if such a thing can be done long-term for aravis cameras. For now I think 1. is a good solution, certainly better than 3. which adds an unnecessary dependency. 

So we're doing roughly the right thing here, but I don't know if it will work with packaging, depending on how `pyproject.toml` is set up. I haven't checked, this is a smell thing, I just haven't always found packaging non-`.py` files in `src` to work reliably so I prefer to embed stuff in `.py` files because, again, it's the most standard thing you can do. 

I'm willing to be overruled on files, however I'd like a different format than csv. We seem to be using yaml everywhere else so if we want to agree on a consistent approach to lookup tables, shall we stick with that?